### PR TITLE
[COST-3484] Process ROS Report

### PR DESCRIPTION
**Pull request by cgoodfred**
## Jira Ticket

[COST-3484](https://issues.redhat.com/browse/COST-3484)

## Description

This change will collect ROS reports during payload processing, send them to S3 for consumption by the ROS team and send a kafka message containing relevant information and the processed files. 

## Testing
Testing for this really has to be done in ephemeral, please reach out if you need help through this part.
1. Checkout Branch
2. Reserve an ephemeral namespace
3. Deploy koku to ephemeral
4. After it has deployed, create a source and send a payload to ingress for an operator report that contains ROS metrics. (if you need help with this part, reach out and I can help)
5. Monitor the processing flow and make sure everything processes appropriately. Check minio and the kafka messages in ephemeral to ensure they are correct.

Sample kafka message, the URL is generated by [an s3 client](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3/client/generate_presigned_url.html#generate-presigned-url).
```
{
  "request_id": "ingress-service-665ff59b4b-jxx2z/naZpbyUQCy-001151",
  "b64_identity": "eyJpZGVudGl0eSI6IHsib3JnX2lkIjogIjEyMzQ1NjciLCAiYWNjb3VudF9udW1iZXIiOiAiMTAwMDEiLCAidHlwZSI6ICJVc2VyIiwgInVzZXIiOiB7InVzZXJuYW1lIjogInVzZXJfZGV2IiwgImVtYWlsIjogInVzZXJfZGV2QGZvby5jb20iLCAiaXNfb3JnX2FkbWluIjogdHJ1ZSwgImFjY2VzcyI6IHt9fX0sImVudGl0bGVtZW50cyI6IHsiY29zdF9tYW5hZ2VtZW50IjogeyJpc19lbnRpdGxlZCI6IHRydWV9fX0=",
  "metadata": {
    "account": "10001",
    "org_id": "1234567",
    "source_id": "c40f7c0a-c96b-4d33-aebe-5a1ccc1c5ad1",
    "cluster_uuid": "023d9b0e-7ca6-481d-b04f-ea606becd54e",
    "cluster_alias": "ocp-ros-testing"
  },
  "files": [
    "http://env-ephemeral-y4ldf0-minio.ephemeral-y4ldf0.svc:9000/ros-eph-s3/org1234567/source%3Dc40f7c0a-c96b-4d33-aebe-5a1ccc1c5ad1/date%3D2023-03-29/008f31ac-67e2-4682-bf87-69ba400584e1_openshift_usage_report.4.csv?AWSAccessKeyId=<key>&Signature=<signature>&Expires=1680267516",
    "http://env-ephemeral-y4ldf0-minio.ephemeral-y4ldf0.svc:9000/ros-eph-s3/org1234567/source%3Dc40f7c0a-c96b-4d33-aebe-5a1ccc1c5ad1/date%3D2023-03-29/008f31ac-67e2-4682-bf87-69ba400584e1_openshift_usage_report.5.csv?AWSAccessKeyId=<key>&Signature=<signature>&Expires=1680267516"
  ],
  "object_keys": [
    "org1234567/source=c40f7c0a-c96b-4d33-aebe-5a1ccc1c5ad1/date=2023-03-29/008f31ac-67e2-4682-bf87-69ba400584e1_openshift_usage_report.4.csv",
    "org1234567/source=c40f7c0a-c96b-4d33-aebe-5a1ccc1c5ad1/date=2023-03-29/008f31ac-67e2-4682-bf87-69ba400584e1_openshift_usage_report.5.csv"
  ]
}
```
Example LOGs from Ephemeral:
When Processing just koku reports:
```
[2023-03-27 12:55:03,561] INFO None {'message': 'No ROS reports to handle for manifest: 1', 'tracing_id': 'ingress-service-6bd86c44d5-8ntq6/eVSWrsjHG2-000407', 'account': '10001', 'org_id': '1234567', 'cluster_id': '023d9b0e-7ca6-481d-b04f-ea606becd54e', 'provider_uuid': 'b97fe3a1-26bd-4489-b8fb-8146a4187897', 'provider_type': 'OCP'}
```

When Processing ROS reports:
```
[2023-03-27 12:56:22,085] INFO None {'message': 'Preparing to upload ROS reports to S3 bucket for manifest: 2.', 'tracing_id': 'ingress-service-6bd86c44d5-8ntq6/eVSWrsjHG2-000438', 'account': '10001', 'org_id': '1234567', 'cluster_id': '023d9b0e-7ca6-481d-b04f-ea606becd54e', 'provider_uuid': 'b97fe3a1-26bd-4489-b8fb-8146a4187897', 'provider_type': 'OCP'}
[2023-03-27 12:56:22,432] INFO None {'message': '1 reports uploaded to S3 for ROS, sending kafka message.', 'tracing_id': 'ingress-service-6bd86c44d5-8ntq6/eVSWrsjHG2-000438', 'account': '10001', 'org_id': '1234567', 'cluster_id': '023d9b0e-7ca6-481d-b04f-ea606becd54e', 'provider_uuid': 'b97fe3a1-26bd-4489-b8fb-8146a4187897', 'provider_type': 'OCP'}
[2023-03-27 12:56:22,437] INFO None kafka message delivered.
```
## Notes

...

### CODE REVIEW

**Comment by djnakabaale** [comment]
Testing results are looking good to me

<details>
<summary>
    Kafka message:
</summary>
    
```
{
    "request_id": "ingress-service-7c7567dbc7-rkn4d/pR1PebQWxU-000757", 
    "b64_identity": "eyJpZGVudGl0eSI6IHsib3JnX2lkIjogIjEyMzQ1NjciLCAiYWNjb3VudF9udW1iZXIiOiAiMTAwMDEiLCAidHlwZSI6ICJVc2VyIiwgInVzZXIiOiB7InVzZXJuYW1lIjogInVzZXJfZGV2IiwgImVtYWlsIjogInVzZXJfZGV2QGZvby5jb20iLCAiaXNfb3JnX2FkbWluIjogdHJ1ZSwgImFjY2VzcyI6IHt9fX0sImVudGl0bGVtZW50cyI6IHsiY29zdF9tYW5hZ2VtZW50IjogeyJpc19lbnRpdGxlZCI6IHRydWV9fX0=", 
    "metadata": {
        "account": "10001", 
        "org_id": "1234567", 
        "source_id": "0e92fa7b-b67f-4b27-bad9-39e3f044531d", 
        "cluster_uuid": "023d9b0e-7ca6-481d-b04f-ea606becd54e", 
        "cluster_alias": "ocp-ros-testing"
    }, 
    "files": ["org1234567/source=0e92fa7b-b67f-4b27-bad9-39e3f044531d/2023-03-27/008f31ac-67e2-4682-bf87-69ba400584e1_openshift_usage_report.4.csv"]
}
``` 
</details>

<details>
<summary>
    Logs from processing ROS reports:
</summary>
    
```
[2023-03-27 15:54:06,718] INFO None {'message': 'Preparing to upload ROS reports to S3 bucket for manifest: 1.', 'tracing_id': 'ingress-service-7c7567dbc7-rkn4d/pR1PebQWxU-000757', 'account': '10001', 'org_id': '1234567', 'cluster_id': '023d9b0e-7ca6-481d-b04f-ea606becd54e', 'provider_uuid': '0e92fa7b-b67f-4b27-bad9-39e3f044531d', 'provider_type': 'OCP'}
[2023-03-27 15:54:07,106] INFO None {'message': '1 reports uploaded to S3 for ROS, sending kafka confirmation.', 'tracing_id': 'ingress-service-7c7567dbc7-rkn4d/pR1PebQWxU-000757', 'account': '10001', 'org_id': '1234567', 'cluster_id': '023d9b0e-7ca6-481d-b04f-ea606becd54e', 'provider_uuid': '0e92fa7b-b67f-4b27-bad9-39e3f044531d', 'provider_type': 'OCP'}
[2023-03-27 15:54:07,111] INFO None kafka message delivered.
``` 
</details>

File: koku/masu/external/kafka_msg_handler.py [hunk 1]
```diff
@@ -329,27 +326,44 @@ def extract_payload(url, request_id, context={}):  # noqa: C901
...
+    ros_processor = ROSReportShipper(
+        context["account"],
+        b64_identity,
+        cluster_id,
+        report_meta["manifest_id"],
+        context["org_id"],
+        report_meta["provider_uuid"],
+        request_id,
+        schema_name,
+        context,
```
[the remark this answers is not in the document]
Comment (cgoodfred) [reply koku/masu/external/kafka_msg_handler.py]: Good suggestion, updated the class to take identity, report meta and context and extract in the init function.

File: koku/masu/external/ros_report_shipper.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+            s3_obj = {"bucket_name": settings.S3_ROS_BUCKET_NAME, "key": upload_key}
+            upload = s3_resource.Object(**s3_obj)
+            upload.upload_fileobj(data, ExtraArgs=extra_args)
+        except (EndpointConnectionError, ClientError) as err:
+            msg = f"Unable to copy data to {upload_key} in bucket {settings.S3_ROS_BUCKET_NAME}.  Reason: {str(err)}"
+            LOG.warning(log_json(self.request_id, msg))
+        return upload_key
+
+    @KAFKA_CONNECTION_ERRORS_COUNTER.count_exceptions()
+    def send_kafka_confirmation(self, msg):
```
[the remark this answers is not in the document]
Comment (cgoodfred) [reply koku/masu/external/ros_report_shipper.py]: True, changed the function name.

**Review by maskarb — DISMISSED** [review]
lgtm! nicely done.

**Review by maskarb — APPROVED** [review]

**Pull request merged**